internal/loadgen/eventhandler/handler.go

```diff
@@ -255,19 +255,38 @@ func (h *Handler) SendBatchesInLoop(ctx context.Context) error {
 		burst: h.config.Limiter.Burst(),
 	}
 
+	// Use a buffer of 1 to avoid blocking the sendErrs channel
+	// if we immediately encounter an error.
+	sendErrs := make(chan error, 1)
+
+	// Send events in batches and when getting an error restart from where
+	// we left.
+	go func() {
+		defer close(sendErrs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				if _, err := h.sendBatches(ctx, &s); err != nil {
+					sendErrs <- err
+					continue
+				}
+				// safeguard `s.sent` so that it doesn't exceed math.MaxInt
+				// but keep the remainder so the next batches know where to start
+				if s.burst > 0 {
+					s.sent = s.sent % s.burst
+				}
+			}
+		}
+	}()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		default:
-			if _, err := h.sendBatches(ctx, &s); err != nil {
-				return err
-			}
-			// safeguard `s.sent` so that it doesn't exceed math.MaxInt
-			// but keep the remainder so the next batches know where to start
-			if s.burst > 0 {
-				s.sent = s.sent % s.burst
-			}
+		case err := <-sendErrs:
```

Comment:
will race happen and accidentally enter this branch when we call close(sendErrs)?

Reply:
no because the channel is closed by the producer above on context cancelation, so at the same time the first case branch in this select is chose. also consider that select has case ordering since 1.15 IIRC, this means that it will enter more likely the first branch if it finds both conditions can be satisfied.
can write a test for it if you think it's something we want to ensure.

Reply:
> select has case ordering since 1.15 IIRC

There is no case ordering. It is non-deterministic, see playground https://go.dev/play/p/Ic9i8MMKMQF

Reply:
when ignoreErrors is true and a signal is received, the sendBatch goroutine in line 264 returns because of <-ctx.Done() and calls line 265 close(sendErrs). then in the main goroutine, both cases in select are ready, and there's a chance that line 291 runs with err == nil. WDYT?

Reply:
^ more like a nit, since this code is not critical and only logs a non-error during shutdown

Reply:
> when ignoreErrors is true 
 
yeah I see I created confusion here, apologies - when IgnoreErrors is true, sendBatch never returns an error, so the if condition is redundant, see `Transport.SendEvents`.

> the sendBatch goroutine in line 264 returns because of <-ctx.Done() 

when the context is canceled, the error may end up in sendErrs, yes - err is not nil, it's context.Canceled, but the gorotuine either takes the first branch of select (and exits) or the message ends up in the channel, and then in the next iteration the channel is closed after return.
we may end up logging a spurious context canceled error, but note that at same time context is canceled, the outer gorotuine with the select at L#286 also returns - so it's very unlikely we hit that log line.
that's also the reason why the channel has a buffer of 1